File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public: 
    bool isValid(string s) {
        stack<char> st;
        for (char c : s) { 
            if (c == '(') 
                st.push(')');
            else if (c == '{')
                st.push('}'); 
            else if (c == '[') 
                st.push(']');
            else {
                if (st.empty() || st.top() != c) // Check if stack is empty before accessing top
                    return false;
                st.pop(); // Pop after comparing
            }
        }
        return st.empty(); // Ensure no unmatched open brackets remain
    }
};




/*
class Solution {
public:
    bool isValid(string s) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);
        
        stack<int> st;
        int n = s.size();

        if(n == 1){
            return false;
        }
        if(n == 0){
            return true;
        }

        for(int i = 0; i < n; i++){
            char ch = s[i];

            if(ch == ')' || ch == '}' || ch == ']'){
                if(st.empty()){
                    return false;
                }
                if(ch == '}' && st.top() != '{')
                    return false;
                if(ch == ')' && st.top() != '(')
                    return false;
                if(ch == ']' && st.top() != '[')
                    return false;
                //else it must be a valid conndition | () or {} or [] |
                st.pop();
            }
            else
            {
                st.push(ch);
            }
        }

        return st.empty()?true : false;
    }
};
*/

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        

        for(char c : s.toCharArray()){
            if(c == ')' || c == '}' || c == ']'){
                if(st.isEmpty()){
                    return false;
                }
                if(st.peek() != '(' && c == ')'){
                    return false;
                }
                if(st.peek() != '{' && c == '}'){
                    return false;
                }
                if(st.peek() != '[' && c == ']'){
                    return false;
                }

                st.pop();

            } else {
                st.push(c);
            }
        }

        return st.isEmpty()?true : false;
    }
}



/*
class Solution {
    public boolean isValid(String s) {
        // Use a deque as a stack to keep track of the opening brackets
        Deque<Character> stack = new ArrayDeque<>();

        for(char c: s.toCharArray()){
            // If the current character is an opening bracket, push it onto the stack
            if(c == '(' || c == '{' || c == '['){
                stack.push(c);
            } else {
                // If the stack is empty or the current closing bracket doesn't match
                // the last opening bracket, the string is not valid
                if(stack.isEmpty() || !match(stack.pop(), c)){
                    return false;
                }
            }
        }
        // If the stack is empty, all brackets were matched correctly
        return stack.isEmpty();
    }

    // Helper method to determine if two brackets are a matching pair
    private boolean match(char left, char right){
        return (left == '(' && right == ')') ||
               (left == '{' && right == '}') ||
               (left == '[' && right == ']');
    }
}
*/